airbyte-integrations/connectors/source-s3/unit_tests/v4/test_cursor.py

```diff
@@ -47,15 +64,282 @@
+            ],
+            id="no_state",
+        ),
+        pytest.param(
+            {
+                "history": {
+                    "file1.txt": "2023-08-01T10:11:12.000000Z",
+                    "file2.txt": "2023-08-01T10:11:12.000000Z",
+                    "file3.txt": "2023-07-31T23:59:59.999999Z",
+                },
+                "_ab_source_file_last_modified": "2023-08-01T10:11:12.000000Z_file2.txt",
+            },
+            [
+                RemoteFile(uri="file1.txt", last_modified=_create_datetime("2023-08-01T10:11:12.000000Z")),
+                RemoteFile(uri="file2.txt", last_modified=_create_datetime("2023-08-01T10:11:12.000000Z")),
+                RemoteFile(uri="file3.txt", last_modified=_create_datetime("2023-07-31T23:59:59.999999Z")),
+            ],
+            [],
+            id="input_state_is_v4_no_new_files",
+        ),
+        pytest.param(
+            {
+                "history": {
+                    "file1.txt": "2023-08-01T10:11:12.000000Z",
+                    "file2.txt": "2023-08-01T10:11:12.000000Z",
+                    "file3.txt": "2023-07-31T23:59:59.999999Z",
+                },
+                "_ab_source_file_last_modified": "2023-08-01T10:11:12.000000Z_file2.txt",
+            },
+            [
+                RemoteFile(uri="file1.txt", last_modified=_create_datetime("2023-08-01T10:11:12.000000Z")),
+                RemoteFile(uri="file2.txt", last_modified=_create_datetime("2023-08-01T10:11:12.000000Z")),
+                RemoteFile(uri="file3.txt", last_modified=_create_datetime("2023-07-31T23:59:59.999999Z")),
+                RemoteFile(uri="file4.txt", last_modified=_create_datetime("2023-08-02T00:00:00.000000Z")),
+            ],
+            [RemoteFile(uri="file4.txt", last_modified=_create_datetime("2023-08-02T00:00:00.000000Z"))],
+            id="input_state_is_v4_with_new_file_later_than_cursor",
+        ),
+        pytest.param(
+            {
+                "history": {
+                    "file1.txt": "2023-08-01T10:11:12.000000Z",
+                    "file2.txt": "2023-08-01T10:11:12.000000Z",
+                    "file3.txt": "2023-07-31T23:59:59.999999Z",
+                },
+                "_ab_source_file_last_modified": "2023-08-01T10:11:12.000000Z_file2.txt",
+            },
+            [
+                RemoteFile(uri="file1.txt", last_modified=_create_datetime("2023-08-01T10:11:12.000000Z")),
+                RemoteFile(uri="file2.txt", last_modified=_create_datetime("2023-08-01T10:11:12.000000Z")),
+                RemoteFile(uri="file3.txt", last_modified=_create_datetime("2023-07-31T23:59:59.999999Z")),
+                RemoteFile(uri="file4.txt", last_modified=_create_datetime("2023-08-01T00:00:00.000000Z")),
+            ],
+            [RemoteFile(uri="file4.txt", last_modified=_create_datetime("2023-08-01T00:00:00.000000Z"))],
+            id="input_state_is_v4_with_new_file_earlier_than_cursor",
```

Comment:
There's one difference between this and the issue - "3. Store a special key in the v4 state message if it was created from a v3 state, v3_min_sync_date (?), whose value is the earliest date in the v3 history. If this key is present, v4 will not sync files before that date."

The purpose of this was to allow us to sync files >1 hour if they weren't already in the history. Did we ultimately decide it wasn't needed?

If we do want it we'll want to add these tests:
- v4 state message + there is one new file, but it's before the `v3_min_sync_date` (as I called it in the issue; this file should not be synced)
- v4 state message + there's one new file and it's between the `v3_min_sync_date` and the earliest file (this should be synced).

Reply:
my bad - you're right. We do need to keep track of `v3_min_sync_date` because v3 removes files from the history based on the time window.